File: OpenBurnBarCore/Sources/OpenBurnBarMedia/VideoDecoderConfigurationPayload.swift
import Foundation

/// Prefix carried inside keyframe `MediaFrame.payload` values when the sender
/// needs to bootstrap a hardware decoder with codec parameter sets.
public struct VideoDecoderConfigurationPayload: Sendable, Equatable {
    public enum Codec: UInt8, Sendable, Equatable {
        case hevc = 1
        case h264 = 2
    }

    public enum DecodeError: Error, Equatable, Sendable {
        case invalidMagic
        case truncated
        case unknownCodec(UInt8)
        case emptyParameterSets
    }

    public static let magic = Data([0x4F, 0x42, 0x56, 0x43, 0x46, 0x47, 0x31]) // OBVCFG1

    public var codec: Codec
    public var parameterSets: [Data]
    public var samplePayload: Data

    public init(codec: Codec, parameterSets: [Data], samplePayload: Data) {
        self.codec = codec
        self.parameterSets = parameterSets
        self.samplePayload = samplePayload
    }

    public func encoded() throws -> Data {
        guard !parameterSets.isEmpty else { throw DecodeError.emptyParameterSets }
        var data = Data()
        data.append(Self.magic)
        data.append(codec.rawValue)
        data.append(UInt8(min(parameterSets.count, Int(UInt8.max))))
        for parameterSet in parameterSets.prefix(Int(UInt8.max)) {
            appendUInt16BigEndian(UInt16(parameterSet.count), to: &data)
            data.append(parameterSet)
        }
        appendUInt32BigEndian(UInt32(samplePayload.count), to: &data)
        data.append(samplePayload)
        return data
    }

    public static func decodeIfPresent(_ data: Data) throws -> VideoDecoderConfigurationPayload? {
        guard data.count >= magic.count else { return nil }
        guard data.prefix(magic.count) == magic else { return nil }

        var offset = magic.count
        guard data.count >= offset + 2 else { throw DecodeError.truncated }
        let codecByte = data[offset]
        offset += 1
        guard let codec = Codec(rawValue: codecByte) else {
            throw DecodeError.unknownCodec(codecByte)
        }
        let count = Int(data[offset])
        offset += 1
        guard count > 0 else { throw DecodeError.emptyParameterSets }

        var parameterSets: [Data] = []
        parameterSets.reserveCapacity(count)
        for _ in 0..<count {
            guard data.count >= offset + 2 else { throw DecodeError.truncated }
            let length = Int(readUInt16BigEndian(from: data, at: offset))
            offset += 2
            guard data.count >= offset + length else { throw DecodeError.truncated }
            parameterSets.append(data.subdata(in: offset..<(offset + length)))
            offset += length
        }

        guard data.count >= offset + 4 else { throw DecodeError.truncated }
        let sampleLength = Int(readUInt32BigEndian(from: data, at: offset))
        offset += 4
        guard data.count >= offset + sampleLength else { throw DecodeError.truncated }
        let samplePayload = data.subdata(in: offset..<(offset + sampleLength))
        return VideoDecoderConfigurationPayload(
            codec: codec,
            parameterSets: parameterSets,
            samplePayload: samplePayload
        )
    }

    private func appendUInt16BigEndian(_ value: UInt16, to data: inout Data) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    private func appendUInt32BigEndian(_ value: UInt32, to data: inout Data) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    private static func readUInt16BigEndian(from data: Data, at offset: Int) -> UInt16 {
        var raw: UInt16 = 0
        _ = withUnsafeMutableBytes(of: &raw) { dest in
            data.copyBytes(to: dest, from: offset..<(offset + 2))
        }
        return UInt16(bigEndian: raw)
    }

    private static func readUInt32BigEndian(from data: Data, at offset: Int) -> UInt32 {
        var raw: UInt32 = 0
        _ = withUnsafeMutableBytes(of: &raw) { dest in
            data.copyBytes(to: dest, from: offset..<(offset + 4))
        }
        return UInt32(bigEndian: raw)
    }
}

File: android/app/src/main/java/com/openburnbar/data/media/VideoDecoderConfigurationPayload.kt
@file:Suppress("MagicNumber", "ThrowsCount", "UnnecessaryParentheses")
// Wire-format NAL/codec byte layout literals; named constants obscure OBVCFG1 parsing.

package com.openburnbar.data.media

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Android port of the Mac/iOS `VideoDecoderConfigurationPayload`.
 *
 * VideoToolbox emits length-prefixed H.264/HEVC NAL units. The Mac sender wraps
 * keyframes with codec parameter sets (`OBVCFG1`) so receivers can bootstrap a
 * hardware decoder. Android `MediaCodec` ByteBuffer input is most reliable with
 * Annex-B start-code delimited access units, so the receive path unwraps this
 * payload and normalizes all length-prefixed NAL units before queueing them.
 */
internal data class VideoDecoderConfigurationPayload(
    val codec: Codec,
    val parameterSets: List<ByteArray>,
    val samplePayload: ByteArray,
) {
    enum class Codec(val rawValue: Byte) {
        HEVC(1),
        H264(2);

        companion object {
            fun from(rawValue: Byte): Codec? = entries.firstOrNull { it.rawValue == rawValue }
        }
    }

    fun encoded(): ByteArray {
        require(parameterSets.isNotEmpty()) { "parameter sets cannot be empty" }
        val payloadSize = MAGIC.size +
            1 +
            1 +
            parameterSets.sumOf { 2 + it.size } +
            4 +
            samplePayload.size
        val buffer = ByteBuffer.allocate(payloadSize).order(ByteOrder.BIG_ENDIAN)
        buffer.put(MAGIC)
        buffer.put(codec.rawValue)
        buffer.put(parameterSets.size.coerceAtMost(255).toByte())
        parameterSets.take(255).forEach { parameterSet ->
            require(parameterSet.size <= UShort.MAX_VALUE.toInt()) { "parameter set too large" }
            buffer.putShort(parameterSet.size.toShort())
            buffer.put(parameterSet)
        }
        buffer.putInt(samplePayload.size)
        buffer.put(samplePayload)
        return buffer.array()
    }

    companion object {
        val MAGIC = byteArrayOf(0x4F, 0x42, 0x56, 0x43, 0x46, 0x47, 0x31) // OBVCFG1

        fun decodeIfPresent(data: ByteArray): VideoDecoderConfigurationPayload? {
            if (data.size < MAGIC.size || !data.startsWith(MAGIC)) return null
            var offset = MAGIC.size
            require(data.size >= offset + 2) { "truncated video decoder config" }
            val codec = Codec.from(data[offset++])
                ?: throw IllegalArgumentException("unknown video decoder config codec")
            val count = data[offset++].toInt() and 0xFF
            require(count > 0) { "empty video decoder parameter sets" }

            val parameterSets = mutableListOf<ByteArray>()
            repeat(count) {
                require(data.size >= offset + 2) { "truncated video decoder parameter set length" }
                val length = readUInt16(data, offset)
                offset += 2
                require(data.size >= offset + length) { "truncated video decoder parameter set" }
                parameterSets += data.copyOfRange(offset, offset + length)
                offset += length
            }

            require(data.size >= offset + 4) { "truncated video decoder sample length" }
            val sampleLength = readInt32(data, offset)
            offset += 4
            require(sampleLength >= 0 && data.size >= offset + sampleLength) {
                "truncated video decoder sample payload"
            }
            return VideoDecoderConfigurationPayload(
                codec = codec,
                parameterSets = parameterSets,
                samplePayload = data.copyOfRange(offset, offset + sampleLength),
            )
        }

        private fun ByteArray.startsWith(prefix: ByteArray): Boolean =
            size >= prefix.size && prefix.indices.all { this[it] == prefix[it] }

        private fun readUInt16(data: ByteArray, offset: Int): Int =
            (data[offset].toInt() and 0xFF shl 8) or
                (data[offset + 1].toInt() and 0xFF)

        private fun readInt32(data: ByteArray, offset: Int): Int =
            (data[offset].toInt() and 0xFF shl 24) or
                (data[offset + 1].toInt() and 0xFF shl 16) or
                (data[offset + 2].toInt() and 0xFF shl 8) or
                (data[offset + 3].toInt() and 0xFF)
    }
}

internal object VideoPayloadNormalizer {
    private val START_CODE = byteArrayOf(0x00, 0x00, 0x00, 0x01)

    fun normalizeForMediaCodec(payload: ByteArray): ByteArray {
        val decoderConfig = VideoDecoderConfigurationPayload.decodeIfPresent(payload)
        if (decoderConfig != null) {
            val normalizedParameterSets = decoderConfig.parameterSets.flatMapToByteArray { withStartCode(it) }
            val normalizedSample = lengthPrefixedNalUnitsToAnnexBOrOriginal(decoderConfig.samplePayload)
            return normalizedParameterSets + normalizedSample
        }
        return lengthPrefixedNalUnitsToAnnexBOrOriginal(payload)
    }

    private fun lengthPrefixedNalUnitsToAnnexBOrOriginal(payload: ByteArray): ByteArray {
        if (payload.hasAnnexBStartCode()) return payload
        val nalUnits = splitLengthPrefixedNalUnits(payload) ?: return payload
        return nalUnits.flatMapToByteArray { withStartCode(it) }
    }

    private fun splitLengthPrefixedNalUnits(payload: ByteArray): List<ByteArray>? {
        if (payload.size < 5) return null
        val units = mutableListOf<ByteArray>()
        var offset = 0
        while (offset < payload.size) {
            if (payload.size - offset < 4) return null
            val length = (payload[offset].toInt() and 0xFF shl 24) or
                (payload[offset + 1].toInt() and 0xFF shl 16) or
                (payload[offset + 2].toInt() and 0xFF shl 8) or
                (payload[offset + 3].toInt() and 0xFF)
            offset += 4
            if (length <= 0 || payload.size - offset < length) return null
            units += payload.copyOfRange(offset, offset + length)
            offset += length
        }
        return units.takeIf { it.isNotEmpty() }
    }

    private fun ByteArray.hasAnnexBStartCode(): Boolean =
        size >= 4 && this[0] == 0.toByte() && this[1] == 0.toByte() &&
            (this[2] == 1.toByte() || this[2] == 0.toByte() && this[3] == 1.toByte())

    private fun withStartCode(nal: ByteArray): ByteArray = START_CODE + nal

    private inline fun List<ByteArray>.flatMapToByteArray(transform: (ByteArray) -> ByteArray): ByteArray {
        val transformed = map(transform)
        val totalSize = transformed.sumOf { it.size }
        val output = ByteArray(totalSize)
        var offset = 0
        transformed.forEach { bytes ->
            bytes.copyInto(output, offset)
            offset += bytes.size
        }
        return output
    }
}
